csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs

```diff
@@ -114,98 +114,614 @@ public IDisposableReadOnlyCollection<DisposableNamedOnnxValue> Run(IReadOnlyColl
+            IReadOnlyCollection<string> outputNames,
+            IReadOnlyCollection<FixedBufferOnnxValue> outputValues)
+        {
+            Run(inputNames, inputValues, outputNames, outputValues, _builtInRunOptions);
+        }
+
+        /// <summary>
+        /// Runs the loaded model for the given inputs and outputs. Uses the given RunOptions for this run.
+        /// 
+        /// Outputs need to be created with correct type and dimension to accept the fetched data.
+        /// </summary>
+        /// <param name="inputNames">Specify a collection of string that indicates the input names. Should match <paramref name="inputValues"/>.</param>
+        /// <param name="inputValues">Specify a collection of <see cref="FixedBufferOnnxValue"/> that indicates the input values.</param>
+        /// <param name="outputNames">Specify a collection of string that indicates the output names. Should match <paramref name="outputValues"/>.</param>
+        /// <param name="outputValues">Specify a collection of <see cref="FixedBufferOnnxValue"/> that indicates the output values.</param>
+        /// <param name="options"></param>
+        public void Run(
+            IReadOnlyCollection<string> inputNames,
+            IReadOnlyCollection<FixedBufferOnnxValue> inputValues,
+            IReadOnlyCollection<string> outputNames,
+            IReadOnlyCollection<FixedBufferOnnxValue> outputValues,
+            RunOptions options)
+        {
+            if (inputNames.Count != inputValues.Count)
+            {
+                throw new ArgumentException($"Length of {nameof(inputNames)} ({inputNames.Count}) must match that of {nameof(inputValues)} ({inputValues.Count}).");
+            }
+            if (outputNames.Count != outputValues.Count)
+            {
+                throw new ArgumentException($"Length of {nameof(outputNames)} ({outputNames.Count}) must match that of {nameof(outputValues)} ({outputValues.Count}).");
+            }
+
+            // prepare inputs
+            string[] inputNamesArray = inputNames as string[] ?? inputNames.ToArray();
+            IntPtr[] inputValuesArray = new IntPtr[inputNames.Count];
+            int inputIndex = 0;
+            foreach (var input in inputValues)
+            {
+                inputValuesArray[inputIndex] = input.Value;
+
+                inputIndex++;
+            }
+
+            // prepare outputs
+            string[] outputNamesArray = outputNames as string[] ?? outputNames.ToArray();
+            IntPtr[] outputValuesArray = new IntPtr[outputNames.Count];
+            int outputIndex = 0;
+            foreach (var output in outputValues)
+            {
+                outputValuesArray[outputIndex] = output.Value;
+
+                // for pre-allocated output, only numberic tensors are supported
+                if (output.OnnxValueType != OnnxValueType.ONNX_TYPE_TENSOR || output.ElementType == TensorElementType.String)
```

Comment:
Both inputs and outputs are of type FixedBufferOnnxValue --why check for string type only for outputs?  If  input is also fixed buffer value (pre-allocated buffers for speedup), then  shouldn't there be an exception if input is string type?

Reply:
I think - FixedBufferOnnxValue can be created from string tensors and the underlying native OrtValue (created only once at object construction) can be used across multiple inference runs. There is no reason to prevent FixedBufferOnnxValue inputs that hold string content, as I see it. In fact, it seems advantageous to support especially when the same string tensor is going to be scored multiple times. This would prevent all the OrtValue creation overhead each time.

Reply:
there are 2 reasons why string tensors should be disallowed as pre-allocated output:

1- there is no performance benefits. the underlying native implementation for pre-allocated output string tensors is using assign operator of `std::string` to a new created value. so even with pre-allocated string tensors the memory used for string is actually not "pre-allocated"

2- this makes inconsistency for C#, if we support pre-allocated string tensors as output. C# string is not a view of native string, there need to be marshaling between native string and managed string. the value of the string tensor will not be updated to the expected output


in short, there is no real pre-allocate for non-numeric tensors (string are always allocated saperately as their length is unknown when allocating), so no need to support them.


For inputs, however, as @hariharans29 mentioned, there is valid requirement so we support them.

Reply:
>it seems advantageous to support especially when the same string tensor is going to be scored multiple times

* **Scenario 1**: A user scores the same string tensor many times, against a model that has only 1 string input.: e.g. {"xxx"}, {"xxx"}, {"xxx"}, {"xxx"} ... {"xxx"}

I assume **scenario 1** is supported using FixedBufferOnnxValue, but there's no point in supporting it because the model will produce the same output for every example. An application can just cache the model score if there's a high frequency string.

*  **Scenario 2**:  A user would like to score 4 different string examples  back to back : e.g.  {"xxx"}, {"yyy"}, {"zzzzz"}, {"xxx"}  


Is  **scenario 2** also supported by using FixedBufferOnnxValue? If **not supported**, then which strings should be removed (minimum number) to make the scenario supportable?

Reply:
If scenario 1 is tweaked so that an app scores multiple models that take the same input ? Would it be beneficial to the overall perf if we did not have to do the string marshaling, create native OrtValue in each model's Run() ?

Reply:
> If scenario 1 is tweaked so that an app scores multiple models that take the same input ? Would it be beneficial to the overall perf if we did not have to do the string marshaling

1. The tweaked scenario (assuming there's some benefit) seems quite uncommon. At least it's not used in any of the ORT examples, UT or tutorials :/

1. More importantly, have a look at the new unit test `TestReusingFixedBufferOnnxValueNonStringTypeMultiInferences()`, at the line **inputs.CopyTo(bufferInput, 0);** 
That looks like the primary usage pattern for FixedBufferOnnxValue() to get a speedup. If that's the main pattern, how would it apply to string types? If there's a different pattern for string-type speedup should be a UT/example for it -- otherwise users won't know about it.